### PR TITLE
Get a list post brief

### PR DESCRIPTION
**Pull request by thuyvy3110**
The requirement is to get a list of all Post Brief #40:
Slug
Title
Brief Description
Time
Tag
View numbers
Comment numbers
Like numbers

### CODE REVIEW

File: src/main/java/com/dawn/jat/illuminati/post/exception/QueryListPostNotFoundException.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.dawn.jat.illuminati.post.exception;
+
+import java.util.Map;
+
+public class QueryListPostNotFoundException extends RuntimeException {
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/exception/QueryListPostNotFoundException.java]: keep it simple here, we dont need to complicate it by using a map

Comment (DeKal) [reply src/main/java/com/dawn/jat/illuminati/post/exception/QueryListPostNotFoundException.java]: Remove Query out of name

File: src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java [hunk 2]
```diff
@@ -1,15 +1,32 @@
...
     }
+
+    /**
+     * Throw Exception when get List Post Not Found.
+     */
+    @ExceptionHandler(QueryListPostNotFoundException.class)
+    public ResponseEntity<Map<String, Object>> exceptionListPost(
+            final QueryListPostNotFoundException exception) {
+        Map<String, Object> errorInfo = new LinkedHashMap<>();
+        errorInfo.put("errorMessage", exception.getMessageFailResp());
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java]: Directly handle the error body here

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 3]
```diff
@@ -38,4 +41,26 @@
...
         return new ResponseEntity<>(postEntity, HttpStatus.OK);
     }
-}
+
+    /**
+     * Gets list of Post Brief.
+     */
+    @GetMapping(value = "/list-post-brief")
+    public ResponseEntity<Object> getListPostBrief() {
+        List<PostEntity> allPost = postService.findAll();
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: We dont need the whole PostEntity here
Change `findAll` to new method `findPostSummary` 
There are 2 solution:
1. Create a new Entity PostBrief, or PostSummary contained only necessary information, it's still connect to collection post. PostService will call PostSummaryRepository to query all post summary
2. Still using the PostRepository, but filter out only necessary information and return back

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 4]
```diff
@@ -38,4 +41,26 @@
         }
         return new ResponseEntity<>(postEntity, HttpStatus.OK);
     }
-}
+
+    /**
+     * Gets list of Post Brief.
+     */
+    @GetMapping(value = "/list-post-brief")
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: `@GetMapping(value = "/all-post/summary")`
change to this for better url 👯‍♂

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 5]
```diff
@@ -38,4 +41,26 @@
...
+    public ResponseEntity<Object> getListPostBrief() {
+        List<PostEntity> allPost = postService.findAll();
+
+        Map<String, Object> body = new LinkedHashMap<>();
+        Boolean success = true;
+
+        if (allPost.isEmpty()) {
+            success = false;
+            body.put("error", "List of Post Brief Not Found");
```
[the remark this answers is not in the document]
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 6]
```diff
@@ -38,4 +41,26 @@
...
+            body.put("success", success);
+            throw new QueryListPostNotFoundException(body);
+        }
+
+        body.put("data", allPost);
+        body.put("success", success);
```
[the remark this answers is not in the document]
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/exception/QueryListPostNotFoundException.java (hunk 1 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/exception/QueryListPostNotFoundException.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java (hunk 4 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: image

Updated

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java (hunk 3 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java (hunk 2 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java]: image

Updated

Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java]: image

Updated

Comment (DeKal) [reply src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java]: I think better keep it in the format
```
{
    error: {
        code: "204",
        message: "Cannot...."    
    },
    success: false
}
```

File: src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java [hunk 7]
```diff
@@ -1,15 +1,31 @@
...
         return new ResponseEntity<>(exception.getMessage(), HttpStatus.NOT_FOUND);
     }
+
+    /**
+     * Throw Exception if Post Summary has No Content.
+     */
+    @ExceptionHandler(PostSummaryNotFoundException.class)
+    public ResponseEntity<ErrorResponse> exceptionPostSummary(Exception exception) {
+        ErrorResponse error = new ErrorResponse();
+        error.setErrorCode(HttpStatus.NO_CONTENT.toString());
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java]: Fix the response to this  to keep it the same with the success one
```
{
    error: {
        code: "204",
        message: "Cannot...."    
    },
    success: false
}
```

File: src/main/java/com/dawn/jat/illuminati/post/exception/ErrorResponse.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.dawn.jat.illuminati.post.exception;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class ErrorResponse {
+    private String errorCode;
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/exception/ErrorResponse.java]: We can use a Map<String,String> for this I think :D

Comment (DeKal) [reply src/main/java/com/dawn/jat/illuminati/post/exception/ErrorResponse.java]: private Map<String,String> error;

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 9]
```diff
@@ -38,4 +37,23 @@
...
+    public ResponseEntity<Object> getAllPostSummary() {
+        List<PostSummaryEntity> allPost = postService.findPostSummary();
+
+        Map<String, Object> body = new LinkedHashMap<>();
+
+        if (allPost.isEmpty()) {
+            throw new PostSummaryNotFoundException("Cannot find post summary");
+        }
+
+        body.put("data", allPost);
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Map<String, Object> body = new LinkedHashMap<>();
 is too far from here

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 10]
```diff
@@ -38,4 +37,23 @@
...
+        body.put("success", true);
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: We should create 2 objects SuccessResponse as oppose to ErrorResponse for this

Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java (hunk 9 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/exception/ErrorResponse.java (hunk 8 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/exception/ErrorResponse.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java (hunk 10 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: image

File: src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java (hunk 7 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionAdvice.java]: Updated

**Pull request merged**